include/RAJA/policy/atomic_builtin.hpp

```diff
@@ -37,389 +44,828 @@ namespace RAJA
-                                        unsigned value)
+
+/*!
+ * Atomic compare and swap using intrinsic
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicCAS(T *acc, T compare, T value)
 {
   __atomic_compare_exchange_n(
-      acc, &compare, value, false, __ATOMIC_ACQ_REL, __ATOMIC_RELAXED);
+      acc, &compare, value, false, __ATOMIC_RELAXED, __ATOMIC_RELAXED);
   return compare;
 }
 
 
-RAJA_DEVICE_HIP
-RAJA_INLINE unsigned long long builtin_atomic_CAS(
-    unsigned long long *acc,
-    unsigned long long compare,
-    unsigned long long value)
+/*!
+ * Atomic addition using intrinsic
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicAdd(T *acc, T value)
 {
-  __atomic_compare_exchange_n(
-      acc, &compare, value, false, __ATOMIC_ACQ_REL, __ATOMIC_RELAXED);
-  return compare;
+  return __atomic_fetch_add(acc, value, __ATOMIC_RELAXED);
 }
 
+
+/*!
+ * Atomic subtraction using intrinsic
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicSub(T *acc, T value)
+{
+  return __atomic_fetch_sub(acc, value, __ATOMIC_RELAXED);
+}
+
+
+/*!
+ * Atomic and using intrinsic
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicAnd(T *acc, T value)
+{
+  return __atomic_fetch_and(acc, value, __ATOMIC_RELAXED);
+}
+
+
+/*!
+ * Atomic or using intrinsic
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicOr(T *acc, T value)
+{
+  return __atomic_fetch_or(acc, value, __ATOMIC_RELAXED);
+}
+
+
+/*!
+ * Atomic xor using intrinsic
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicXor(T *acc, T value)
+{
+  return __atomic_fetch_xor(acc, value, __ATOMIC_RELAXED);
+}
+
+
 #endif  // RAJA_COMPILER_MSVC
 
 
+/*!
+ * Atomics implemented using reinterpret cast
+ */
+
+
+/*!
+ * Alias for determining the integral type of the same size as the given type
+ */
 template <typename T>
-RAJA_DEVICE_HIP RAJA_INLINE
-    typename std::enable_if<sizeof(T) == sizeof(unsigned), T>::type
-    builtin_atomic_load(T *acc)
+using builtin_useReinterpret_t = typename builtin_useReinterpret<T>::type;
+
+
+/*!
+ * Atomic load using reinterpret cast
+ */
+template <typename T,
+          std::enable_if_t<builtin_useReinterpret<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicLoad(T *acc)
 {
-  return RAJA::util::reinterp_A_as_B<unsigned, T>(
-      builtin_atomic_load((unsigned *)acc));
+  return RAJA::util::reinterp_A_as_B<builtin_useReinterpret_t<T>, T>(
```

Comment:
Would it be nicer to use local alias `using R = builtin_useReinterpret_t<T>`?

Reply:
Yup. Done.